runtime/src/main/java/org/corfudb/runtime/LogReplicationListener.java

```diff
@@ -0,0 +1,182 @@
+
+    // Timestamp at which the client performed a full sync.  Any updates below this timestamp must be ignored.
+    // At the time of subscription, a full sync cannot be performed if LR Snapshot Sync is in progress.  Full Sync is
+    // performed when this ongoing snapshot sync completes.  The listener, however, can get updates before this full
+    // sync.  So we need to maintain this timestamp and ignore any updates below it.
+    @Getter
+    private final AtomicLong clientFullSyncTimestamp = new AtomicLong(Address.NON_ADDRESS);
+
+    private final CorfuStore corfuStore;
+    private final String namespace;
+
+    /**
+     * Special LogReplication listener which a client creates to receive ordered updates for replicated data.
+     * @param corfuStore Corfu Store used on the client
+     * @param namespace Namespace of the client's tables
+     */
+    public LogReplicationListener(CorfuStore corfuStore, @Nonnull String namespace) {
+        this.corfuStore = corfuStore;
+        this.namespace = namespace;
+    }
+
+    /**
+     * This is an internal method of this abstract listener and not exposed to clients.
+     *
+     * @param results is a map of stream UUID -> list of entries of this stream.
+     */
+    public final void onNext(CorfuStreamEntries results) {
+
+        // If this update came before the client's full sync timestamp, ignore it.
+        if (results.getTimestamp().getSequence() <= clientFullSyncTimestamp.get()) {
+            return;
+        }
+
+        Set<String> tableNames =
+                results.getEntries().keySet().stream().map(schema -> schema.getTableName()).collect(Collectors.toSet());
+
+        if (tableNames.contains(REPLICATION_STATUS_TABLE)) {
+            Preconditions.checkState(results.getEntries().keySet().size() == 1,
+                "Replication Status Table Update received with other tables");
+            processReplicationStatusUpdate(results);
+            return;
+        }
+
+        // Data Updates
+        if (clientFullSyncPending.get()) {
+            // If the listener started when snapshot sync was ongoing, ignore all data updates until it ends.  When
+            // it ends, the client will perform a full sync and build a consistent state containing these updates.
+            return;
+        }
+
+        if (snapshotSyncInProgress.get()) {
+            processUpdatesInSnapshotSync(results);
+        } else {
+            processUpdatesInLogEntrySync(results);
+        }
+    }
+
+    private void processReplicationStatusUpdate(CorfuStreamEntries results) {
+        Map<TableSchema, List<CorfuStreamEntry>> entries = results.getEntries();
+
+        List<CorfuStreamEntry> replicationStatusTableEntries =
+            entries.entrySet().stream().filter(e -> e.getKey().getTableName().equals(REPLICATION_STATUS_TABLE))
+            .map(Map.Entry::getValue)
+            .findFirst()
+            .get();
```

Comment:
I thought by findFirst we will only get the first CorfuStreamEntry? or we are assuming StatusTable update sent to listener will be single entry. otherwise I remember here we basically want to collect all the CorfuStreamEntry belongs to StatusTable, should it be something like this?
```
            .map(Map.Entry::getValue)
            .flatMap(List::stream)
            .collect(Collectors.toList());
```

Reply:
Okay I see it now, thanks!